lib/internal/util/debuglog.js

```diff
@@ -115,7 +147,159 @@ function debuglog(set, cb) {
+}
+
+/**
+ * @type {Record<string, SafeMap>}
+ */
+let tracesStores;
+
+/**
+ * @typedef {(logLabel: string, traceLabel?: string) => void} TimerStart
+ */
+
+/**
+ * @typedef {(label: string, traceLabel?: string) => void} TimerEnd
+ */
+
+/**
+ * @typedef {(label: string, traceLabel?: string, args?: any[]) => void} TimerLog
+ */
+
+/**
+ * Debuglog with time fns and support for trace
+ * @param {string} set
+ * @param {(startTimer: TimerStart, endTimer: TimerEnd, logTimer: TimerLog) => void} cb
+ * @returns {{startTimer: TimerStart, endTimer: TimerEnd, logTimer: TimerLog}}
+ */
+function debugWithTimer(set, cb) {
+  set = StringPrototypeToUpperCase(set);
+
+  if (tracesStores === undefined) {
+    tracesStores = { __proto__: null };
+  }
+
+  /**
+   * @type {import('internal/util/trace_timer').LogImpl}
+   */
+  function logImpl(label, timeFormatted, args) {
+    const pid = process.pid;
+    const colors = { colors: lazyUtilColors().shouldColorize(process.stderr) };
+    const coloredPID = inspect(pid, colors);
+
+    if (args === undefined)
+      process.stderr.write(format('%s %s %s: %s\n', set, coloredPID, label, timeFormatted));
+    else
+      process.stderr.write(
+        format(
+          '%s %s %s: %s\n',
+          set,
+          coloredPID,
+          label,
+          timeFormatted,
+          ...new SafeArrayIterator(args),
+        ),
+      );
+  }
+
+  const kTraceCategory = buildCategory(StringPrototypeToLowerCase(set));
+  /**
+   * @type {import('internal/util/trace_timer').LogImpl}
+   */
+  let selectedLogImpl;
+  let categoryEnabled = false;
+
+  /**
+   * @type {TimerStart}
+   */
+  function internalStartTimer(logLabel, traceLabel) {
+    if (!categoryEnabled && !isTraceCategoryEnabled(kTraceCategory)) {
+      return;
+    }
+
+    lazyTraceTimer().time(tracesStores[set], kTraceCategory, 'debuglog.time', logLabel, traceLabel);
+  }
+
+  /**
+   * @type {TimerEnd}
+   */
+  function internalEndTimer(logLabel, traceLabel) {
+    if (!categoryEnabled && !isTraceCategoryEnabled(kTraceCategory)) {
+      return;
+    }
+
+    lazyTraceTimer()
+      .timeEnd(tracesStores[set], kTraceCategory, 'debuglog.timeEnd', selectedLogImpl, logLabel, traceLabel);
+  }
+
+  /**
+   * @type {TimerLog}
+   */
+  function internalLogTimer(logLabel, traceLabel, args) {
+    if (!categoryEnabled && !isTraceCategoryEnabled(kTraceCategory)) {
+      return;
+    }
+
+    lazyTraceTimer()
+      .timeLog(tracesStores[set], kTraceCategory, 'debuglog.timeLog', selectedLogImpl, logLabel, traceLabel, args);
+  }
+
+  function init() {
+    if (tracesStores[set] === undefined) {
+      tracesStores[set] = new SafeMap();
+    }
+    emitWarningIfNeeded(set);
+    categoryEnabled = testEnabled(set);
+    selectedLogImpl = categoryEnabled ? logImpl : noop;
+  }
+
+  /**
+   * @type {TimerStart}
+   */
+  const startTimer = (logLabel, traceLabel) => {
+    init();
+
+    if (cb)
+      cb(internalStartTimer, internalEndTimer, internalLogTimer);
```

Comment:
It should do something even cheaper - detect whether the environment variables are enabled, if not, do `cb(noop, noop, noop)` - and also once the `cb` is invoked, don't invoke it again, the same should also be done for the two other methods. This is what we do for `debuglog` (`debuglogImpl()` returns the `noop` constants which gets re-assigned as the variable in the caller via the callback).

Reply:
I can't do that, `trace_events` can be enabled at runtime, so there's no way to avoid using the original implementation here and I always need to check if the category is enabled.